File: source/com/barrybecker4/simulation/predprey/PredPreySimulator.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.simulation.predprey;

import com.barrybecker4.common.math.function.CountFunction;
import com.barrybecker4.common.math.function.Function;
import com.barrybecker4.simulation.common.ui.Simulator;
import com.barrybecker4.simulation.common.ui.SimulatorApplet;
import com.barrybecker4.simulation.common.ui.SimulatorOptionsDialog;
import com.barrybecker4.simulation.graphing1.GraphOptionsDialog;
import com.barrybecker4.simulation.predprey.creatures.Foxes;
import com.barrybecker4.simulation.predprey.creatures.Population;
import com.barrybecker4.simulation.predprey.creatures.Rabbits;
import com.barrybecker4.simulation.predprey.options.DynamicOptions;
import com.barrybecker4.ui.renderers.MultipleFunctionRenderer;
import com.barrybecker4.ui.util.GUIUtil;

import javax.swing.JPanel;
import java.awt.Color;
import java.awt.Graphics;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * Simulates foxes (predators) and rabbits (prey) in the wild.
 *
 * @author Barry Becker
 */
public class PredPreySimulator extends Simulator {

    private MultipleFunctionRenderer graph;
    private long iteration;

    private Rabbits rabbits;
    private Foxes foxes;

    private CountFunction rabbitFunction;
    private CountFunction foxFunction;

    private DynamicOptions options_;

    /** Constructor */
    private PredPreySimulator() {
        super("Predator Prey Simulation");

        foxes = new Foxes();
        rabbits = new Rabbits();

        initGraph();
    }

    public List<Population> getCreatures() {
        List<Population> creatures = new ArrayList<Population>();
        creatures.add(foxes);
        creatures.add(rabbits);
        return creatures;
    }

    @Override
    protected void reset() {

        initGraph();
        options_.reset();
    }

    @Override
    protected double getInitialTimeStep() {
        return 1.0;
    }

    @Override
    public double timeStep() {
        iteration++;

        foxes.setPopulation( foxes.getPopulation() * foxes.birthRate
                           - foxes.getPopulation() * foxes.deathRate / rabbits.getPopulation());
        //                   - lions.getPopulation() * foxes.deathRate * foxes.getPopulation());
        rabbits.setPopulation( rabbits.getPopulation() * rabbits.birthRate
                             - foxes.getPopulation() * rabbits.deathRate * rabbits.getPopulation());
        //                    - lions.getPopulation() * rabbits.deathRate * rabbits.getPopulation());

        rabbitFunction.addValue(iteration, rabbits.getPopulation());
        foxFunction.addValue(iteration, foxes.getPopulation());

        options_.update();

        return timeStep_;
    }

    private  void initGraph() {
        iteration = 0;

        rabbits.reset();
        foxes.reset();

        rabbitFunction = new CountFunction(Rabbits.INITIAL_NUM_RABBITS);
        foxFunction = new CountFunction(Foxes.INITIAL_NUM_FOXES);

        List<Function> functions = new LinkedList<Function>();
        functions.add(rabbitFunction);
        functions.add(foxFunction);
        //functions.add(lionFunction);

        List<Color> lineColors = new LinkedList<Color>();
        lineColors.add(Rabbits.COLOR);
        lineColors.add(Foxes.COLOR);
        //lineColors.add(Lions.COLOR);

        graph = new MultipleFunctionRenderer(functions, lineColors);
    }


    @Override
    public JPanel createDynamicControls() {
        options_ = new DynamicOptions(this);
        return options_;
    }


    @Override
    protected SimulatorOptionsDialog createOptionsDialog() {
         return new GraphOptionsDialog(frame, this );
    }

    @Override
    public void paint( Graphics g ) {
        graph.setSize(getWidth(), getHeight());
        graph.paint(g);
    }

    public static void main( String[] args ) {
        final PredPreySimulator sim = new PredPreySimulator();

        sim.setPaused(true);
        GUIUtil.showApplet(new SimulatorApplet(sim));
    }
}

File: scala-source/com/barrybecker4/simulation/predprey/PredPreySimulator.scala
// Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.predprey

import com.barrybecker4.math.function.CountFunction
import com.barrybecker4.math.function.Function
import com.barrybecker4.simulation.common.ui.{Simulator, SimulatorOptionsDialog}
import com.barrybecker4.simulation.predprey.creatures.Foxes
import com.barrybecker4.simulation.predprey.creatures.Population
import com.barrybecker4.simulation.predprey.creatures.Rabbits
import com.barrybecker4.simulation.predprey.options.DynamicOptions
import com.barrybecker4.ui.renderers.MultipleFunctionRenderer

import javax.swing.JPanel
import java.awt.Color
import java.awt.Graphics

import scala.compiletime.uninitialized


class PredPreySimulator() extends Simulator("Predator Prey Simulation") {

  private var foxes = new Foxes
  private var rabbits = new Rabbits

  private var graph: MultipleFunctionRenderer = uninitialized
  private var iteration = 0L
  private var rabbitFunction: CountFunction = uninitialized
  private var foxFunction: CountFunction = uninitialized
  private var options: Option[DynamicOptions] = None
  initGraph()

  def getCreatures: Seq[Population] = Seq(foxes, rabbits)

  override protected def reset(): Unit = {
    initGraph()
    options.foreach(_.reset())
  }

  override protected def getInitialTimeStep = 1.0

  override def timeStep: Double = {
    iteration += 1
    stepPopulations()
    recordGraphAndUi()
    tStep
  }

  private def stepPopulations(): Unit =
    val f = foxes.getPopulation
    val r = rabbits.getPopulation
    foxes.setPopulation(
      PredPreyDynamics.nextFoxCount(f, r, foxes.birthRate, foxes.deathRate)
    )
    rabbits.setPopulation(
      PredPreyDynamics.nextRabbitCount(r, foxes.getPopulation, rabbits.birthRate, rabbits.deathRate)
    )

  private def recordGraphAndUi(): Unit =
    rabbitFunction.addValue(iteration.toDouble, rabbits.getPopulation.toDouble)
    foxFunction.addValue(iteration.toDouble, foxes.getPopulation.toDouble)
    options.foreach(_.update())

  private def initGraph(): Unit = {
    iteration = 0
    rabbits.reset()
    foxes.reset()
    rabbitFunction = new CountFunction(Rabbits.INITIAL_NUM_RABBITS)
    foxFunction = new CountFunction(Foxes.INITIAL_NUM_FOXES)
    val functions = List[Function](rabbitFunction, foxFunction)
    val lineColors = List[Color](Rabbits.COLOR, Foxes.COLOR)
    graph = new MultipleFunctionRenderer(functions, Some(lineColors))
  }

  override def createDynamicControls: JPanel = {
    val panel = new DynamicOptions(this)
    options = Some(panel)
    panel
  }

  override protected def createOptionsDialog = new SimulatorOptionsDialog(frame, this)

  override def paint(g: Graphics): Unit = {
    graph.setSize(getWidth, getHeight)
    graph.paint(g)
  }
}
